src/main/java/org/rocstreaming/roctoolkit/NativeObject.java

```diff
@@ -1,18 +1,101 @@
 package org.rocstreaming.roctoolkit;
 
-class NativeObject {
+import static java.util.concurrent.TimeUnit.SECONDS;
 
-    private long ptr;
+/**
+ * A <code>NativeObject</code> represents an underlying native roc object.
+ */
+class NativeObject implements AutoCloseable {
 
-    NativeObject() {
-        ptr = 0L;
+    /**
+     * Maximum time to wait (in milliseconds) for joining {@link AutoCloseThread}.
+     */
+    private final static long MAX_JOIN_TIMEOUT_MS = SECONDS.toMillis(20L);
+
+    /**
+     * <code>NativeObject</code> finalizer thread.
+     */
+    private final static AutoCloseThread thread = AutoCloseThread.getInstance();
+
+    /**
+     *  Underlying roc object native pointer.
+     */
+    private final long ptr;
+
+    /**
+     *  Destructor method.
+     */
+    private final Destructor destructor;
+
+    /**
+     *  Reference to {@link NativeObjectReference}.
+     */
+    private final NativeObjectReference resource;
```

Comment:
`resource` is reference to the same NativeObject, right? If took some time to me to get this. Maybe call it `referenceToThis` or something like this?